### ISSUE TITLE
Canceling `JSHost.ImportAsync` fails because `dynamic_import` in `invoke-js.ts` is not cancelable

### ISSUE THREAD

**Issue by nikonthethird** [dotnet/runtime#80028]
### Description

# TLDR:
The cancellation token given to `JSHost.ImportAsync` does not work because the underlying JS function called by `ImportAsync` does not return a `Promise` that is "controllable" (outfitted with a special symbol).

```c#
// Repro here: https://github.com/nikonthethird/ImportAsyncError, or paste this code into an InitializeAsync  of a page in Blazor WASM:
var cts = new CancellationTokenSource();
cts.Cancel();
await JSHost.ImportAsync("hello", "world.js", cts.Token);
```
The code above fails with an exception because [`dynamic_import`](https://github.com/dotnet/runtime/blob/8ac3832812f9bc37fba2a66ac8987ca66d381d85/src/mono/wasm/runtime/invoke-js.ts#L177) is not controllable (the promise generated by the function is not wrapped using `wrap_as_cancelable_promise` like for example [`http_wasm_fetch`](https://github.com/dotnet/runtime/blob/8ac3832812f9bc37fba2a66ac8987ca66d381d85/src/mono/wasm/runtime/http.ts#L52) is).

# Description

After upgrading to .NET 7, an application of mine has been throwing the following error messages in the log:

```text
"Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer"]: Unhandled exception rendering component: "Error: Assert failed: Promise is not controllable"
at System.Runtime.InteropServices.JavaScript.CancelablePromise._CancelPromise(IntPtr )
   at System.Runtime.InteropServices.JavaScript.CancelablePromise.CancelPromise(Task )
   at System.Net.Http.BrowserHttpInterop.<>c__DisplayClass13_0`1[[System.Int32, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<CancelationHelper>b__0()
   at System.Threading.CancellationToken.<>c.<Register>b__12_0(Object )
   at System.Threading.CancellationTokenSource.Invoke(Delegate , Object , CancellationTokenSource )
   at System.Threading.CancellationTokenSource.CallbackNode.<>c.<ExecuteCallback>b__9_0(Object )
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext , ContextCallback , Object )
--- End of stack trace from previous location ---
```
This is what I discovered:

* [`JSHost.ImportAsync`](https://github.com/dotnet/runtime/blob/8ac3832812f9bc37fba2a66ac8987ca66d381d85/src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSHostImplementation.cs#L272) calls `JavaScriptImports.DynamicImport`, then uses the `CancelationHelper` function to [call `CancelablePromise.CancelPromise`](https://github.com/dotnet/runtime/blob/8ac3832812f9bc37fba2a66ac8987ca66d381d85/src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSHostImplementation.cs#L288) when the cancellation token fires.
* [`CancelPromise`](https://github.com/dotnet/runtime/blob/8ac3832812f9bc37fba2a66ac8987ca66d381d85/src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/CancelablePromise.cs#L14) calls [`mono_wasm_cancel_promise`](https://github.com/dotnet/runtime/blob/6251d496e307ed39f1e600cd036106cc1bd7e340/src/mono/wasm/runtime/cancelable-promise.ts#L26) which throws the error above [when the promise is not controllable](https://github.com/dotnet/runtime/blob/6251d496e307ed39f1e600cd036106cc1bd7e340/src/mono/wasm/runtime/promise-controller.ts#L73), meaning when it does not have the promise control symbol.
* However, the promise returned by [`JavaScriptImports.DynamicImport`](https://github.com/dotnet/runtime/blob/8aa77a952a285dcd318cf8e6c99ba6f99cec897f/src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Interop/JavaScriptImports.Generated.cs#L46) is directly from [the async function `dynamic_import`](https://github.com/dotnet/runtime/blob/fda0b358709e4b971b2cf546dff898f2ccaceb2c/src/mono/wasm/runtime/invoke-js.ts#L320), which is **NOT** wrapped with [`wrap_as_cancelable_promise`](https://github.com/dotnet/runtime/blob/fda0b358709e4b971b2cf546dff898f2ccaceb2c/src/mono/wasm/runtime/cancelable-promise.ts#L19) like [all](https://github.com/dotnet/runtime/blob/fda0b358709e4b971b2cf546dff898f2ccaceb2c/src/mono/wasm/runtime/http.ts#L52) [the](https://github.com/dotnet/runtime/blob/fda0b358709e4b971b2cf546dff898f2ccaceb2c/src/mono/wasm/runtime/http.ts#L83) [others](https://github.com/dotnet/runtime/blob/fda0b358709e4b971b2cf546dff898f2ccaceb2c/src/mono/wasm/runtime/http.ts#L106) that use `CancelationHelper` are.

Since .NET 7 uses the `[JSImport]` source generator internally, I suspect the exceptions I am experiencing also come from some internal `JSHost.ImportAsync` calls.

### Reproduction Steps

```razor
@page "/"
@using System.Runtime.InteropServices.JavaScript

<h1>Hello, world!</h1>

@if (error is not null) {
    <p>@error</p>
}

@code {
    private String? error;

    protected override async Task OnInitializedAsync() {
        var cts = new CancellationTokenSource();
        cts.Cancel();

        try {
            await JSHost.ImportAsync("hello", "world.js", cts.Token);
        } catch (Exception ex) {
            error = ex.ToString();
        }
        
    }
}
```

A reproduction can [also be found here](https://github.com/nikonthethird/ImportAsyncError).

### Expected behavior

The import should be cancelled.

### Actual behavior

Error: Assert failed: Promise is not controllable

### Regression?

Yes, no problem with my code in .NET 6.
Note that I do not even use `JSHost.ImportAsync` in my code directly, but .NET 7 itself uses the new `[JSImport]` source generated, and this is where I expect the error to occur.

### Known Workarounds

None so far.

### Configuration

.NET 7.0.1
Ubuntu 20.04 / Firefox
Windows 10 22H2 / Chromium
x64
It happens in all cases, I don't know why I even provided this info...

**Comment by pavelsavara** [comment]
thanks @nikonthethird , your explanation is perfect :)
